firebaseai/LiveAudioExample/Audio/AudioController.swift

```diff
@@ -0,0 +1,256 @@
+  private var routeTask: Task<Void, Never>?
+
+  /// Port types that are considered "headphones" for our use-case.
+  ///
+  /// More specifically, airpods are considered bluetooth ports instead of headphones, so
+  /// this array is necessary.
+  private let headphonePortTypes: [AVAudioSession.Port] = [
+    .headphones,
+    .bluetoothA2DP,
+    .bluetoothLE,
+    .bluetoothHFP,
+  ]
+
+  private let modelInputFormat: AVAudioFormat
+  private let modelOutputFormat: AVAudioFormat
+
+  private var stopped = false
+
+  public init() async throws {
+    let session = AVAudioSession.sharedInstance()
+    try session.setCategory(
+      .playAndRecord,
+      mode: .voiceChat,
+      options: [.defaultToSpeaker, .allowBluetooth, .duckOthers,
+                .interruptSpokenAudioAndMixWithOthers, .allowBluetoothA2DP]
+    )
+    try session.setPreferredIOBufferDuration(0.01)
+    try session.setActive(true)
+
+    guard let modelInputFormat = AVAudioFormat(
+      commonFormat: .pcmFormatInt16,
+      sampleRate: 16000,
+      channels: 1,
+      interleaved: false
+    ) else {
+      fatalError("Failed to create model input format")
+    }
+
+    guard let modelOutputFormat = AVAudioFormat(
+      commonFormat: .pcmFormatInt16,
+      sampleRate: 24000,
+      channels: 1,
+      interleaved: true
+    ) else {
+      fatalError("Failed to create model output format")
+    }
+
+    self.modelInputFormat = modelInputFormat
+    self.modelOutputFormat = modelOutputFormat
+
+    let (processedData, dataQueue) = AsyncStream<AVAudioPCMBuffer>.makeStream()
+    microphoneData = processedData
+    microphoneDataQueue = dataQueue
+
+    listenForRouteChange()
+  }
+
+  deinit {
+    stopped = true
+    listenTask?.cancel()
+    // audio engine needs to be stopped before disconnecting nodes
+    audioEngine?.pause()
+    audioEngine?.stop()
+    if let audioEngine {
+      do {
+        // the VP IO leaves behind artifacts, so we need to disable it to properly clean up
+        if audioEngine.inputNode.isVoiceProcessingEnabled {
+          try audioEngine.inputNode.setVoiceProcessingEnabled(false)
+        }
+      } catch {
+        print("Failed to disable voice processing: \(error.localizedDescription)")
+      }
+    }
+    microphone?.stop()
+    audioPlayer?.stop()
+    microphoneDataQueue.finish()
+    routeTask?.cancel()
+  }
+
+  /// Kicks off audio processing, and returns a stream of recorded microphone audio data.
+  public func listenToMic() throws -> AsyncStream<AVAudioPCMBuffer> {
+    spawnAudioProcessingThread()
+    return microphoneData
+  }
+
+  /// Permanently stop all audio processing.
+  ///
+  /// To start again, create a new instance of ``AudioController``.
+  public func stop() {
+    stopped = true
+    stopListeningAndPlayback()
+    microphoneDataQueue.finish()
+    routeTask?.cancel()
+  }
+
+  /// Queues audio for playback.
+  public func playAudio(audio: Data) {
+    audioPlayer?.play(audio)
```

Comment:
When running on the iOS Simulator, the model interrupts itself when the audio answers are played. One way to work around this is to disable audio output on the Simulator:

```
    #if !targetEnvironment(simulator)
    audioPlayer?.play(audio)
    #endif
```